### PR TITLE
Align PETSc mpi comm with comm of PETSc Vec type, and expose VecNest type

### PR DESCRIPTION
**Pull request by luca-heltai**

### CODE REVIEW

**Comment by luca-heltai** [comment]
@drwells, here as well we have a static MPI_Comm. Let's get all this batch merged as is, and address these static objects in subsequent PR.

**Review by stefanozampini — APPROVED** [review]

File: include/deal.II/lac/petsc_block_vector.h
Comment (drwells) [inline include/deal.II/lac/petsc_block_vector.h:158]:
```suggestion
       * This method assigns the PETSc Vec to the instance of the class.
```

File: include/deal.II/lac/petsc_block_vector.h
Comment (drwells) [inline include/deal.II/lac/petsc_block_vector.h:162]: Would it be possible to use a constructor instead of this function?

File: include/deal.II/lac/petsc_block_vector.h
Comment (drwells) [inline include/deal.II/lac/petsc_block_vector.h:321]: we should set this up in the ctor

File: include/deal.II/lac/petsc_vector_base.h
Comment (drwells) [inline include/deal.II/lac/petsc_vector_base.h:334]:
```suggestion
     * This method assigns the PETSc Vec to the instance of the class.
     * This is particularly useful when performing PETSc to deal.II operations
     * since it allows one to reuse the VectorBase and the PETSc Vec
     * without incurring in memory copies.
```
or use the ctor.

**Review by drwells — CHANGES_REQUESTED** [review]
I think its possible to use the constructor instead of `assign_petsc_vector()`. If its not then lets go with it.

File: source/lac/petsc_matrix_free.cc
Comment (stefanozampini) [reply source/lac/petsc_matrix_free.cc:74]: I think it is better to use `MPI_COMM_WORLD` to be consistent with the rest of the code that uses `MPI_` communicators.

File: include/deal.II/lac/petsc_block_vector.h
Comment (stefanozampini) [reply include/deal.II/lac/petsc_block_vector.h:162]: The point here is that you want to keep the same deal.ii object but associate a different PETSc `Vec`

File: include/deal.II/lac/petsc_vector_base.h
Comment (stefanozampini) [reply include/deal.II/lac/petsc_vector_base.h:334]: again, the constructor will create a new deal.II object. here we want to use the same deal.ii object

File: include/deal.II/lac/petsc_block_vector.h
Comment (luca-heltai) [reply include/deal.II/lac/petsc_block_vector.h:162]: Ps: the version with the constructor is already there (if not in this PR, in a subsequent PR)

File: source/lac/petsc_matrix_free.cc
Comment (luca-heltai) [reply source/lac/petsc_matrix_free.cc:74]: So, which way do we go? `MPI_COMM_WORLD` or `PETSC_COMM_WORLD`?

Comment (drwells) [reply source/lac/petsc_matrix_free.cc:74]: Sorry for the slow response - in principle we should always use `PETSC_COMM_WORLD` here since PETSc may not be set up to use `MPI_COMM_WORLD`, but since we use `MPI_COMM_WORLD` all over the place we should make that change in one place later.

File: include/deal.II/lac/petsc_block_vector.h
Comment (drwells) [reply include/deal.II/lac/petsc_block_vector.h:166]: I'd rather use standard C++ idioms here instead of a new member function. It may be wordier but it will make this consistent with the rest of the library (e.g., if we ever did something similar for direct manipulation of Trilinos objects).

In particular, I'd prefer to create a new object and then swap:
```cpp
PETScWrappers::Vector v1;
// set up v1

Vec v;
// set up v
PETScWrappers::Vector v2(v);
swap(v1, v2);
```
that way we don't have two functions which do the same thing (the ctor and the assign function). Are you two OK with doing it this way instead?

Comment (stefanozampini) [reply include/deal.II/lac/petsc_block_vector.h:166]: I'm ok with using `swap` but I have to think about the implications and the proper implementation. Can we postpone this change? I will open an MR later for this

Comment (stefanozampini) [reply include/deal.II/lac/petsc_block_vector.h:166]: @drwells Ping. Can we merge this in and continue integrating the PETSc additions? I agree with your comments and remove the `assign` method, replacing with `swap` in another PR. @luca-heltai

Comment (stefanozampini) [reply include/deal.II/lac/petsc_block_vector.h:166]: https://github.com/dealii/dealii/issues/14656

File: include/deal.II/lac/petsc_block_vector.h
Comment (bangerth) [inline include/deal.II/lac/petsc_block_vector.h:519]: I don't think this works as expected with the `static`. Why not do this:
```suggestion
      MPI_Comm        pcomm =
        PetscObjectComm(reinterpret_cast<PetscObject>(petsc_nest_vector));
      if (pcomm != MPI_COMM_NULL)
        return pcomm;
      else
        return PETSC_COMM_SELF;
```
But one can also ask whether it is not a failure case if the call to `PetscObjectComm` returns `MPI_COMM_NULL`, and just error out in that case with an assertion.

File: include/deal.II/lac/petsc_matrix_base.h
Comment (bangerth) [inline include/deal.II/lac/petsc_matrix_base.h:1641]: Same as above

File: include/deal.II/lac/petsc_vector_base.h
Comment (bangerth) [inline include/deal.II/lac/petsc_vector_base.h:1164]: Same here again

File: source/lac/petsc_matrix_free.cc
Comment (bangerth) [reply source/lac/petsc_matrix_free.cc:91]: I think we'll need to make a sweep over this issue after these patches are merged. It's now all inconsistent.

File: source/lac/petsc_parallel_block_sparse_matrix.cc
Comment (bangerth) [inline source/lac/petsc_parallel_block_sparse_matrix.cc:182]: And here

File: source/lac/petsc_vector_base.cc
Comment (bangerth) [inline source/lac/petsc_vector_base.cc:158]: Can you elaborate what the TODO refers to? Right now you're probably the only one who actually knows what it means :-)

File: source/lac/petsc_vector_base.cc
Comment (bangerth) [inline source/lac/petsc_vector_base.cc:183]: same here

**Review by bangerth — COMMENTED** [review]
As before, thanks a lot for working on this!

**Review by drwells — APPROVED** [review]
Looks good (sorry it took me some time to get back to this) - lets address Wolfgang's comments and merge.

File: include/deal.II/lac/petsc_block_vector.h
Comment (stefanozampini) [reply include/deal.II/lac/petsc_block_vector.h:519]: It needs `static` because the return argument is a reference to an `MPI_Comm` (I haven't changed it). Your code will compile with warnings, while the one with `static` is ok. Anyway, we will remove the reference from the return argument in a follow-up PR.

Comment (stefanozampini) [reply include/deal.II/lac/petsc_block_vector.h:519]: regarding the `MPI_COMM_NULL` comment: If the PETSc object is `NULL`, then `PetscObjectComm` returns `MPI_COMM_NULL`. It seemed to me that you use SELF when an object is not yet set up to be used in deal.II, that's why it is returning self. That was the rationale, I can change the behavior if you prefer.

File: include/deal.II/lac/petsc_block_vector.h
Comment (stefanozampini) [reply include/deal.II/lac/petsc_block_vector.h:251]: The base class method is not virtual.

**Comment by stefanozampini** [comment]
@bangerth Thanks for taking a look. I think I have answered your points and added notes for follow-up PRs. If so, then let's merge this in and get to the juicy PRs with nonlinear solvers.

File: source/lac/petsc_vector_base.cc
Comment (stefanozampini) [reply source/lac/petsc_vector_base.cc:158]: Added a note here https://github.com/dealii/dealii/issues/14656

Comment (bangerth) [reply source/lac/petsc_vector_base.cc:158]: Can you put your comment above into a comment in the code? This way all of the information needed to address the TODO is in the code.

File: include/deal.II/lac/petsc_block_vector.h
Comment (bangerth) [reply include/deal.II/lac/petsc_block_vector.h:251]: I should have been explicit: Since you're overriding functionality, *should* the function in the base class be `virtual`?

Comment (stefanozampini) [reply include/deal.II/lac/petsc_block_vector.h:251]: Well, I'm not sure. There are explicit calls to `BaseClass::collect_sizes` also for the trilinos case. As usual with C++, is almost always a matter of taste in deciding which feature of the language to adopt.

File: include/deal.II/lac/petsc_block_vector.h
Comment (stefanozampini) [reply include/deal.II/lac/petsc_block_vector.h:519]: This change will be part of the follow-up PRs, so I guess we can consider this discussion resolved.

File: source/lac/petsc_vector_base.cc
Comment (stefanozampini) [reply source/lac/petsc_vector_base.cc:158]: will do

Comment (stefanozampini) [reply source/lac/petsc_vector_base.cc:158]: @bangerth I have added support here https://github.com/dealii/dealii/pull/14481/commits/acf141cc565cb6b4bdc34f56098fbba2e2f18b0c

File: source/lac/petsc_vector_base.cc
Comment (stefanozampini) [reply source/lac/petsc_vector_base.cc:183]: see above

File: include/deal.II/lac/petsc_block_vector.h
Comment (stefanozampini) [reply include/deal.II/lac/petsc_block_vector.h:166]: @bangerth I have renamed with`reinit` here https://github.com/dealii/dealii/pull/14481/commits/757f08692cd4996f064546764b66c6d11c31ecc0

File: include/deal.II/lac/petsc_block_vector.h
Comment (stefanozampini) [reply include/deal.II/lac/petsc_block_vector.h:528]: @bangerth Here you have the relevant commit https://github.com/dealii/dealii/pull/14481/commits/0db14d853d927f7dbb45918d71ebbc0c33dd10a9

File: include/deal.II/lac/petsc_block_vector.h
Comment (stefanozampini) [reply include/deal.II/lac/petsc_block_vector.h:134]: @bangerth documentation improved here https://github.com/dealii/dealii/pull/14481/commits/d76f474c345dcee0b12eb28f64fe1ac890b9665f

**Comment by stefanozampini** [comment]
@bangerth all your points have been addressed with follow-up commits

**Comment by stefanozampini** [comment]
I have merged https://github.com/dealii/dealii/pull/14686 here

File: include/deal.II/lac/petsc_block_vector.h
Comment (bangerth) [reply include/deal.II/lac/petsc_block_vector.h:134]: @stefanozampini Did you forget to push this commit to this branch? It doesn't show up as part of the pull request...

Comment (stefanozampini) [reply include/deal.II/lac/petsc_block_vector.h:134]: I will open a PR later for that, otherwise it will be messier than it is now

**Comment by stefanozampini** [comment]
It seems the CI failure is independent of this PR. Can this be merged or we should trigger a rebuild?

**Comment by drwells** [comment]
I can reproduce the test failure. The last few lines of output are
```
11973: DEAL:0:PETScWrappers::MPI::Vector::test_nvector_view_unwrap OK
11973: DEAL:0:PETScWrappers::MPI::Vector::test_get_vector_id OK
11973: DEAL:0:PETScWrappers::MPI::Vector::test_clone OK
11973: DEAL:0:PETScWrappers::MPI::Vector::test_destroy OK
11973: 
11973: sundials/n_vector.debug: RUN failed. ------ Additional output on stdout/stderr:
11973: 
11973: terminate called after throwing an instance of '(anonymous namespace)::NVectorTestError'
11973:   what():  
11973: --------------------------------------------------------
11973: An error occurred in line <374> of file </not_backed_up/drwells/Documents/Code/CPP/dealii/tests/sundials/n_vector.cc> in function
11973:     void test_get_communicator() [with VectorType = dealii::PETScWrappers::MPI::Vector; typename std::enable_if<(! dealii::is_serial_vector<T>::value), int>::type <anonymous> = 0]
11973: The violated condition was: 
11973:     *static_cast<MPI_Comm *>(N_VGetCommunicator(n_vector)) == MPI_COMM_WORLD
11973: Additional information: 
11973:     The internal N_Vector implementation didn't pass a test.
11973: --------------------------------------------------------
```
i.e., it looks like changing the communicator broke this test.

**Comment by stefanozampini** [comment]
@drwells I have fixed the test, since it does not make sense to compare for the same handle (e.g. using `com1 == com2`). The test now uses `MPI_Comm_compare` since PETSc duplicates the user communicator.

**Comment by drwells** [comment]
Yes, thanks for the fix!

**Review by drwells — APPROVED** [review]

**Comment by stefanozampini** [comment]
@bangerth This can be merged if you have no objections

**Comment by bangerth** [comment]
Thanks as always, @stefanozampini !

**Pull request merged**
